features/config/service/impl/src/main/java/org/opennms/features/config/service/impl/AbstractCmJaxbConfigDao.java

```diff
@@ -0,0 +1,158 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+/**
+ * <p>Abstract AbstractCmJaxbConfigDao class.</p>
+ *
+ * @param <ENTITY_CLASS> Configuration class
+ * @version $Id: $
+ */
+public abstract class AbstractCmJaxbConfigDao<ENTITY_CLASS> {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractCmJaxbConfigDao.class);
+
+    @Autowired
+    private ConfigurationManagerService configurationManagerService;
+
+    private Class<ENTITY_CLASS> entityClass;
+    private String description;
+    private final Collection<Consumer<ENTITY_CLASS>> onReloadCausedChangeCallbacks = new ArrayList<>();
+    private ConcurrentHashMap<String, ENTITY_CLASS> lastKnownEntityMap = new ConcurrentHashMap<>();
+
+    /**
+     * ConfigName use for the Config and Schema
+     *
+     * @return ConfigName
+     */
+    abstract protected String getConfigName();
+
+
+    /**
+     * The default configId when getConfig without passing configId
+     *
+     * @return configId
+     */
+    abstract protected String getDefaultConfigId();
+
+    /**
+     * <p>Constructor for AbstractJaxbConfigDao.</p>
+     *
+     * @param entityClass a {@link java.lang.Class} object.
+     * @param description a {@link java.lang.String} object.
+     */
+    public AbstractCmJaxbConfigDao(final Class<ENTITY_CLASS> entityClass, final String description) {
+        super();
+        this.entityClass = entityClass;
+        this.description = description;
+    }
+
+    /**
+     * loadConfig from CM. If the config is already stored in cache, it will also trigger reload callback
+     *
+     * @param configId
+     * @return ConfigObject
+     * @throws IOException
+     */
+    protected ENTITY_CLASS loadConfig(final String configId) throws IOException {
+        long startTime = System.currentTimeMillis();
+
+        LOG.debug("Loading {} configuration from {}", description, configId);
+        Optional<ENTITY_CLASS> configOptional = configurationManagerService.getConfiguration(this.getConfigName(), configId, entityClass);
+        if (configOptional.isEmpty()) {
+            throw new RuntimeException("NOT_FOUND: configName=" + this.getConfigName() + " configId=" + configId);
+        }
+        ENTITY_CLASS config = configOptional.get();
+        long endTime = System.currentTimeMillis();
+        LOG.info("Loaded {} in {} ms", getDescription(), (endTime - startTime));
+
+        // If the config is not load in the first time, we will trigger the callbacks for this change
+        ENTITY_CLASS lastKnownEntity = lastKnownEntityMap.get(configId);
+        if (lastKnownEntity != null) {
+            synchronized (onReloadCausedChangeCallbacks) {
+                if (!onReloadCausedChangeCallbacks.isEmpty()) {
+                    LOG.debug("Calling onReloaded callbacks");
+                    try {
+                        onReloadCausedChangeCallbacks.forEach(c -> c.accept(config));
```

Comment:
the loop stops at the first failure. I wonder if we shouldn't try each callback even after one failed?

Reply:
added TODO comment to consider when working on PE-13